### ISSUE TITLE
Feature Request: Single-click NetVM Disable/Enable

### ISSUE THREAD

**Issue by straussmaus** [QubesOS/qubes-issues#4834]
### Qubes OS version:

R4.0

### Affected component(s):

Qubes Manager

---

### Steps to reproduce the behavior:

Many time there are network issues that make me stop and restart a VM, maybe maybe VMs.  If there are many VMs connecting to a VM I need to strart I must open the settings for each one individually, change the NetVM, click OK, Wait a few years, then restart the VM.  Then I must do the opposite and add the NetVM back.

This is really bad when the VM I restart is sys-whonix because I have maybe 15 VMs that connect to sys-whonix.

### Expected behavior:

The feature request is to add a column to the QubesVM and in this column there would be a button.  Clicking it would turn it clear and this means the VM is not connected to it's NetVM.  Click it again and it will turn green and attach to it's NetVM.

### Actual behavior:

There is no quick way to do this now.

### General notes:

This is not a complaint just a request because it would save me about 4 hours (really!) every week.  If you do not have strong Internet and you have many VMs then you know what I mean.

---

**Comment by marmarek** [comment]
Ideally this manual operation shouldn't be needed at all, it should all happen automatically.
In practice, there is some preliminary work already done. You can shutdown a qube from within (execute `sudo poweroff` there) to bypass `qvm-shutdown` refusal to shutdown used qube. Then, when you start that qube again, all relevant qubes should automatically re-connect to it. The last step is not very stable yet. But if you'd like to try, please report back here how it works for you.

Also, `sys-whonix` problems most of the times could also be solved with restarting just tor inside (there is a menu entry for that).

**Comment by andrewdavidwong** [comment]
> Ideally this manual operation shouldn't be needed at all, it should all happen automatically.
> In practice, there is some preliminary work already done. You can shutdown a qube from within (execute `sudo poweroff` there) to bypass `qvm-shutdown` refusal to shutdown used qube. Then, when you start that qube again, all relevant qubes should automatically re-connect to it. The last step is not very stable yet. But if you'd like to try, please report back here how it works for you.
> 
> Also, `sys-whonix` problems most of the times could also be solved with restarting just tor inside (there is a menu entry for that).

Closing as "won't do." If anyone has a new reason for why this should be done, please leave a comment, and we'll be happy to take another look. Thank you.

**Comment by z64z** [comment]
@straussmaus Fastest way I found: click Q in icon in top right (Qubes Domains) > select your netvm > pause it > click Q again > select netvm again > use the kill option (only displayed for paused VMs) after it gets killed restart it from the top left Q menu (applications) by e.g. opening a terminal in it. (Or start it using the Qubes Manager.)

**Comment by straussmaus** [comment]
@z64z Yes that's faster but it's still veeeeeeeeery sllllllow compared to my request.  Even more slow when you have to do it to 5 or 6 VMs every time.

@marmarek You said "You can shutdown a qube from within (execute sudo poweroff there) to bypass qvm-shutdown refusal to shutdown used qube" but that does not address my request.  Using qvm-shutdown was not the alternative.  Actually that would take much more time than the way I said I do it now because I would have to change desktops each time, sometimes open a terminal in VMs that do not have one open.  So your suggestion takes more time than the original problem.

Also it is about 0.01% of the time that the *Net/ProxyVM* has a terminal window open and that is the one that needs to be rebooted so often.  So really I do not see why you are connecting your suggestions to my request?  Very different things.

@andrewdavidwong I hope this makes a "reason for why this should be done" more clear.

**Comment by straussmaus** [comment]
@marmarek 

>  You can shutdown a qube from within

I did not reply to this.  You have a good solution for a quick shutdown for users who 

1 - Are using one Qube

2 - Are on the command line

But if you have 5 or 6 open and you have to restart all of their NetVMs individually, you have to click the settings for each one, set to no net, shut down each net, open the settings for each again, and reset to the original netvm - and don't make a mistake!

This feature would save many hours per week.  @andrewdavidwong I hope this explains why this is a good "should do" not a "won't do." :)

**Comment by straussmaus** [comment]
I've probably spend 15 hours of my life on this since opening this issue.  I can't be the only person who would benefit from a Net toggle in the Qubes Manager?

**Comment by andrewdavidwong** [comment]
> I've probably spend 15 hours of my life on this since opening this issue. I can't be the only person who would benefit from a Net toggle in the Qubes Manager?

I don't recall ever having to spend much time on this. It sounds like you're probably making things harder than necessary.

1. Remember that you can switch the NetVM of `sys-firewall` to `none`, then restart `sys-net`. This avoids having to restart any of the VMs connected to `sys-firewall`.
2. Remember that you don't have to restart `sys-whonix` just to restart Tor, and usually don't even have to restart Tor. You just need a new Tor *circuit*, which is easily done by opening Nyx in `sys-whonix` (look for the menu entry), then simply pressing <kbd>n</kbd>.

**Comment by straussmaus** [comment]
Thank you for this.  Neither situation applies to me.  I don't think I have ever changed the sys-firewall NetVM, except when setting up Qubes for the first time.

There are times when it is necessary to restart sys-whonix (or at least more comforting), such as when switching routers, but that is rare.

So these situations aren't the ones eating up half of my life. :)

But if no on else is feeling the same way (which I really don't understand) then there is no need to prioritize this.  Thank you again for your suggestions.

**Comment by donob4n** [comment]
@straussmaus I'm not sure if this is exactly what you want but when https://github.com/QubesOS/qubes-manager/pull/273 is merged you could change template and netVM wich just a right and left click (also you could set the NetVM to None).

**Comment by straussmaus** [comment]
@donob4n 

Yes, that will take a little more user effort (and less space) than my feature request.  It might(?) be better than what I pictured, and I'm happy to know I'm not the only person who wants this feature.

I will try to remember to close or modify this issue when that feature comes online.  Thank you.

**Comment by straussmaus** [comment]
@donob4n I think this is merged?  Will it appear in 4.0.x or do I have to reinstall 4.1?

**Comment by donob4n** [comment]
It is not merged yet https://github.com/QubesOS/qubes-manager/pull/273

If it is something highly valuable for you can install it (although you must understand that installing anything in dom0 is not recommended and you should only do if you understand exactly what are you doing). Just clone my repository, switch to branch 'template_menu' and run "make ui" for generate the ui_* files (maybe you need to install some dependencies for this).

After this you only have to copy the 'qubesmanager' folder to '/usr/lib/python3.5/site-packages/qubesmanager' on dom0.

**Comment by blatchard** [comment]
@donob4n  The fastest way i have found to resolve this is (as you say, you don't usually have a terminal open for your netvm) is to just right click the blue qube in the system tray, start a terminal from there (which you can do in both 4 and 4.1, and just type "sudo shutdown now".  None of my other vm's are shutdown like this, and my proxys, be them wireguard or openvpn, or anything else i have chained seems to work just fine after only the net vm is restarted (add StartQube to your netVM menu from applications). 

Would doing this not resolve your downtime?

As a side note a lot of my issues with NetVM are due to an issue that i will  add if i find it, to do with random allocation of device ordering in netvm (wls6 or wls7) and often because of nm-applet playing up.

**Comment by donob4n** [comment]
@blatchard with https://github.com/QubesOS/qubes-manager/pull/281/commits/60dbeabfb44e5c9535462393a48e898d9b657b12 you could just click restart button from qube mananger and It will be restarted regardless it has connected vm's.

It should be included in v4.1.

**Comment by icequbes1** [comment]
While this issue requests a single-click solution, I've had a small script that I've used to perform VM tasks quickly with a keyboard shortcut. Purpose was to avoid having to use the mouse or remember command lines.

```
# full menu-based prompt for vm and operation
./vmop.sh

# prompt for operation given a vm
./vmop.sh sys-net

# prompt for operation on a vm determined from the active window
./vmop.sh [ACTIVEWINDOW]

# or just do something on a vm determined from active window
./vmop.sh [ACTIVEWINDOW] toggle-network
```

Various things it does: toggle-network, shutdown vm, open vm console, open terminal, clear vm clipboard, open settings, send a key [to an HVM]

Putting the script somewhere in dom0 and configured an xfce keyboard shortcut to run an op on [ACTIVEWINDOW].

So if I configure ctrl+alt+x as an xfce4 keyboard shortcut with command being `vmop.sh [ACTIVEWINDOW] clear-clipboard`, now copying and pasting more securely is as "simple" as ctrl+c, ctrl+shift+c, ctrl+shift+v, ctrl+v, ctrl+shift+x

I doubt this meets any Qubes quality standards but just providing what's worked for me as an example.

[vmop.sh.txt](https://github.com/QubesOS/qubes-issues/files/6643280/vmop.sh.txt)

**Comment by unman** [comment]
On Sat, Jun 12, 2021 at 09:41:46PM -0700, icequbes1 wrote:
> While this issue requests a single-click solution, I've had a small script that I've used to perform VM tasks quickly with a keyboard shortcut. Purpose was to avoid having to use the mouse or remember command lines.
> 
> ```
> # full menu-based prompt for vm and operation
> ./vmop.sh
> 
> # prompt for operation given a vm
> ./vmop.sh sys-net
> 
> # prompt for operation on a vm determined from the active window
> ./vmop.sh [ACTIVEWINDOW]
> 
> # or just do something on a vm determined from active window
> ./vmop.sh [ACTIVEWINDOW] toggle-network
> ```
> 
> Various things it does: toggle-network, shutdown vm, open vm console, open terminal, clear vm clipboard, open settings, send a key [to an HVM]
> 
> Putting the script somewhere in dom0 and configured an xfce keyboard shortcut to run an op on [ACTIVEWINDOW].
> 
> So if I configure ctrl+alt+x as an xfce4 keyboard shortcut with command being `vmop.sh [ACTIVEWINDOW] clear-clipboard`, now copying and pasting more securely is as "simple" as ctrl+c, ctrl+shift+c, ctrl+shift+v, ctrl+v, ctrl+shift+x
> 
> I doubt this meets any Qubes quality standards but just providing what's worked for me as an example.
> 
> [vmop.sh.txt](https://github.com/QubesOS/qubes-issues/files/6643280/vmop.sh.txt)

I'll take a look at the script after - I find that many feature requests
can be resolved with trivial dom0 scripting. We should gather these
together.
